File: implementations/node-express/src/middleware/injection.js
/**
 * Vulnerable injection implementations
 */

import { execSync, exec } from 'child_process';
import { db } from '../database.js';

/**
 * Search products with SQL injection vulnerability
 *
 * VULNERABILITY V06: Direct string concatenation in SQL query
 * Example exploits:
 * - ' OR '1'='1
 * - ' UNION SELECT id, username, email, password_hash, ssn, credit_card, secret_note, role, 1, 1 FROM users --
 */
export function searchProductsVulnerable(searchTerm) {
  // VULNERABLE: String concatenation in SQL
  const query = `SELECT * FROM products WHERE name LIKE '%${searchTerm}%' OR description LIKE '%${searchTerm}%'`;

  try {
    const products = db.prepare(query).all();
    return products;
  } catch (error) {
    // Return error details (also a vulnerability - information disclosure)
    return { error: error.message, query: query };
  }
}

/**
 * Ping a host with command injection vulnerability
 *
 * VULNERABILITY V07: Direct command execution with user input
 * Example exploits:
 * - 127.0.0.1; cat /etc/passwd
 * - 127.0.0.1 && whoami
 * - $(cat /etc/passwd)
 */
export function pingHostVulnerable(host) {
  // VULNERABLE: Direct command execution with user input
  const command = `ping -c 1 ${host}`;

  try {
    const result = execSync(command, {
      timeout: 10000,
      encoding: 'utf-8',
      shell: true  // VULNERABLE: shell=true with user input
    });

    return {
      success: true,
      command: command,  // VULNERABILITY: Exposing executed command
      stdout: result,
      stderr: ''
    };
  } catch (error) {
    return {
      success: false,
      command: command,
      stdout: error.stdout || '',
      stderr: error.stderr || error.message
    };
  }
}

/**
 * DNS lookup with command injection vulnerability
 *
 * VULNERABILITY V07: Another command injection vector
 * Example exploits:
 * - google.com; id
 * - google.com && cat /etc/shadow
 */
export function dnsLookupVulnerable(domain) {
  const command = `nslookup ${domain}`;

  try {
    const result = execSync(command, {
      timeout: 10000,
      encoding: 'utf-8',
      shell: true
    });

    return {
      success: true,
      domain: domain,
      output: result
    };
  } catch (error) {
    return {
      success: false,
      domain: domain,
      error: error.message
    };
  }
}

File: implementations/python-fastapi/app/vulnerabilities/injection.py
"""Vulnerable injection implementations."""

import subprocess
import sqlite3
from typing import Any

from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import text

from app.models import Product


async def search_products_vulnerable(
    db: AsyncSession,
    search_term: str,
) -> list[dict[str, Any]]:
    """
    Search products with SQL injection vulnerability.

    VULNERABILITY V06: Direct string concatenation in SQL query
    Example exploit: ' OR '1'='1' --
    Example exploit: ' UNION SELECT id, username, email, password_hash, ssn, credit_card, secret_note, role FROM users --
    """
    # VULNERABLE: String concatenation in SQL
    query = f"SELECT * FROM products WHERE name LIKE '%{search_term}%' OR description LIKE '%{search_term}%'"

    result = await db.execute(text(query))
    rows = result.fetchall()

    # Convert to dict - column order matches model definition
    products = []
    for row in rows:
        products.append({
            "id": row[0],
            "name": row[1],
            "description": row[2],
            "price": row[3],
            "stock": row[4],
            "category": row[5],
            "is_active": row[6],
            "internal_notes": row[7],
            "supplier_cost": row[8],
            "created_at": row[9],
        })

    return products


def ping_host_vulnerable(host: str) -> dict[str, Any]:
    """
    Ping a host with command injection vulnerability.

    VULNERABILITY V07: Direct command execution with user input
    Example exploit: 127.0.0.1; cat /etc/passwd
    Example exploit: 127.0.0.1 && whoami
    Example exploit: $(cat /etc/passwd)
    """
    # VULNERABLE: Direct command execution with user input
    command = f"ping -c 1 {host}"

    try:
        result = subprocess.run(
            command,
            shell=True,  # VULNERABLE: shell=True with user input
            capture_output=True,
            text=True,
            timeout=10
        )
        return {
            "success": result.returncode == 0,
            "command": command,  # VULNERABILITY: Exposing executed command
            "stdout": result.stdout,
            "stderr": result.stderr,
        }
    except subprocess.TimeoutExpired:
        return {
            "success": False,
            "error": "Command timed out"
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }


def dns_lookup_vulnerable(domain: str) -> dict[str, Any]:
    """
    DNS lookup with command injection vulnerability.

    VULNERABILITY V07: Another command injection vector
    """
    command = f"nslookup {domain}"

    try:
        result = subprocess.run(
            command,
            shell=True,
            capture_output=True,
            text=True,
            timeout=10
        )
        return {
            "success": result.returncode == 0,
            "domain": domain,
            "output": result.stdout,
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }
